### PR TITLE
Fix(dhcp_provisioner): Support custom `node_type_keys` as a list of dicts.

### PR DESCRIPTION
**Pull request by philippebureau**
## Change Summary

Add filter to the creation of node type list to match new AVD 4.x syntax of node type keys (dict of dict -> list of dict)

## Related Issue(s)

Fixes aristanetworks/avd-internal#154

This was preventing the use of custom node type keys with the new AVD 4.0 syntax

## Component(s) name

`arista.avd.dhcp_provisioner`

## Proposed changes
add | map(attribute='key')

## How to test
create custom node types and run the role
see molecule test for example

### Repository Checklist

- [ x ] My code has been rebased from devel before I start
- [ x ] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ x ] My change requires a change to the documentation and documentation have been updated accordingly.
- [ x ] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/molecule/dhcp_configuration/inventory/group_vars/DC1_FABRIC.yml [hunk 1]
```diff
@@ -177,3 +184,67 @@ bfd_multihop:
...
+    type: l2leaf
+    connected_endpoints: true
+    mlag_support: true
+    network_services:
+      l2: true
+    underlay_router: false
+    uplink_type: port-channel
+
+  - key: custom_l2leaf
+    type: customl_2leaf
```
Comment (danieltudares) [inline ansible_collections/arista/avd/molecule/dhcp_configuration/inventory/group_vars/DC1_FABRIC.yml]: fix typo

Comment (philippebureau) [reply ansible_collections/arista/avd/molecule/dhcp_configuration/inventory/group_vars/DC1_FABRIC.yml]: fixed

**Comment by philippebureau** [comment]
after discussing internally, to prevent braking old syntax "| arista.avd.convert_dicts('key')" was added to convert the old syntax so it works with both.
Thank you to @gmuloc for the recommendation

File: ansible_collections/arista/avd/molecule/dhcp_provisioning/inventory/group_vars/DC1_L2LEAFS.yml [hunk 2]
```diff
@@ -1 +1,65 @@
 type: l2leaf
+
+# Custom node types with old syntax
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/dhcp_provisioning/inventory/group_vars/DC1_L2LEAFS.yml]:
```suggestion
# Custom node types with AVD 3.x syntax using a dict of dicts
```

File: ansible_collections/arista/avd/molecule/dhcp_provisioning/inventory/group_vars/DC1_FABRIC.yml [hunk 3]
```diff
@@ -175,3 +185,67 @@ bfd_multihop:
   interval: 1200
   min_rx: 1200
   multiplier: 3
+
+# Custom node types
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/dhcp_provisioning/inventory/group_vars/DC1_FABRIC.yml]: Hmm given that the role takes only the first device in the template to derive the node_type_keys (yeah thats not great we agree) I don't think everything is being tested as we would expect here.
We probably need to create separate fabrics for the default behavior (that was tested before and is not now), the new syntax node_type_keys as you have here and the AVD 3.8 syntax node_type_keys.

I guess lets write 3 tests fabric / modify the playbook to run at three levels?

File: ansible_collections/arista/avd/molecule/dhcp_configuration/inventory/group_vars/DC1_CUSTOM_L2LEAFS.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
+type: custom_l2leaf
+
+# Custom node types with old syntax
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/dhcp_configuration/inventory/group_vars/DC1_CUSTOM_L2LEAFS.yml]:
```suggestion
# Custom node types with AVD 3.8 syntax
```

File: ansible_collections/arista/avd/molecule/dhcp_provisioning/inventory/group_vars/DC1_FABRIC.yml (hunk 3 above)
Comment (philippebureau) [reply ansible_collections/arista/avd/molecule/dhcp_provisioning/inventory/group_vars/DC1_FABRIC.yml]: I will work on the 3 fabrics test

**Review by danieltudares — APPROVED** [review]
LGTM!

**Review by gmuloc — DISMISSED** [review]
LGTM - thanks @philippebureau for taking the time to fix this!

File: ansible_collections/arista/avd/molecule/dhcp_provisioning/molecule.yml [hunk 5]
```diff
@@ -1,29 +1,24 @@
...
-    - prepare
     - converge
   cleanup_sequence:
     - destroy
 dependency:
   name: galaxy
   enabled: false
 driver:
-  name: docker
+  name: default
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/dhcp_provisioning/molecule.yml:19]: Why do we need these changes to the molecule scenario?

Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/dhcp_provisioning/molecule.yml:19]: it is more that we don't need docker for offline mode and so we are triggering extra hooks in the scenario that are not needed. The other scenario `dhcp_configruation` (and yes we can argue that the names are confusing) for online mode still requires docker containers.

**Review by ClausHolbechArista — APPROVED** [review]

**Pull request merged**